dbms/programs/client/Client.cpp

```diff
@@ -87,12 +91,79 @@ namespace ErrorCodes
     extern const int CLIENT_OUTPUT_FORMAT_SPECIFIED;
 }
 
-
 class Client : public Poco::Util::Application
 {
 public:
     Client() {}
 
+    void init_suggestions(Completion::TSTNode &node)
+    {
+        QUERYPART *qP = queryParts;
+        while (qP->name) {
+            node.add_word(qP->name);
+            qP++;
+        }
+
+        std::vector<Block> blocks;
+
+        //preload all functions
+        sendQuery("SELECT name FROM system.functions", blocks);
+        for (const auto &block : blocks) {
+            size_t size = block.rows();
+            for (size_t i = size; i--;) {
+                node.add_word(strdup(block.getByName("name").column->getDataAt(i).toString().c_str()));
+            }
+        }
+        blocks.clear();
+
+        int limit = config().getInt("suggestion_limit", suggestion_limit);
+
+        //preload limited amount of dbs
+        auto *query = new char[128];
+        sprintf(query, "SELECT name FROM system.databases ORDER BY name LIMIT %d", limit);
+        sendQuery(query, blocks);
+        for (const auto &block : blocks) {
+                size_t size = block.rows();
+                for (size_t i = size; i--;) {
+                    node.add_word(strdup(block.getByName("name").column->getDataAt(i).toString().c_str()));
+                }
+        }
+        blocks.clear();
+        delete query;
+
+        //preload limited amount of tables and their columns
+        query = new char[128];
+        sprintf(
+                query,
+                "SELECT table,name FROM system.columns WHERE table IN (SELECT name FROM system.tables ORDER BY name"
+                " LIMIT %d)",
+                limit
+        );
+        sendQuery(query, blocks);
+        for (const auto &block : blocks) {
+            ColumnWithTypeAndName tableNameColumn = block.getByName("table");
+            ColumnWithTypeAndName columnNameColumn = block.getByName("name");
+                size_t size = block.rows();
+                for (size_t i = size; i--;) {
+                    const char *tableName = strdup(tableNameColumn.column->getDataAt(i).toString().c_str());
+                    const char *columnName = strdup(columnNameColumn.column->getDataAt(i).toString().c_str());
+                    auto *implodedName = new char[strlen(tableName)+strlen(columnName)+2];
+                    sprintf(implodedName, "%s.%s", tableName, columnName);
+                    if (!node.has(tableName))  {
+                        node.add_word(tableName);
+                    }
+                    if (!node.has(columnName)) {
+                        node.add_word(columnName);
+                    }
+                    if (!node.has(implodedName)) {
+                        node.add_word(implodedName);
+                    }
+                    delete implodedName;
+            }
+        }
+        delete query;
```

Comment:
Also dictionary autocomplete is VERY helpful. 

Check that: 
https://github.com/filimonov/chc/blob/8bb7b4a1a7917ff5cb75f84e7b76091463f23ecc/autocomplete.go#L196

Reply:
@filimonov Added dictionary name and attribute name as part of completion. Can add whole dictGetT() template as well. But I think there will be problems with complex keys.